server/blockchain/calypso/calypso.go

```diff
@@ -455,32 +499,87 @@ func (c calypsoCommand) revealSecret(snap store.Snapshot, step execution.Step) e
 			name, smcKey)
 	}
 
-	secret, err := snap.Get(name)
+	secret, err := getSecret(snap, name)
 	if err != nil {
 		return xerrors.Errorf("failed to get secret '%s': %v", name, err)
 	}
 
-	// Generate authorization hash
-	h := sha256.New()
-	h.Write(smcKey)
-	h.Write(secret)
-	h.Write(clientPubKey)
-	hash := h.Sum(nil)
+	accessToken := computeAccessToken(smcKey, secret, clientPubKey)
 
-	// TODO: this should be done according to the store specification at the
-	// top of this file, but the prefixing logic has not yet been implemented.
-	err = snap.Set(hash, hash)
+	if hasSecretAccess(snap, accessToken) {
+		return xerrors.Errorf("secret '%s' was already revealed", name)
+	}
+
+	err = setSecretAccess(snap, accessToken, clientPubKey)
 	if err != nil {
 		return xerrors.Errorf("failed to persist secret reveal: %v", err)
 	}
 
+	err = insertAuditLog(snap, name, accessToken)
+	if err != nil {
+		return xerrors.Errorf("failed to persist secret audit log: %v", err)
+	}
+
 	dela.Logger.Info().
 		Str("contract", ContractName).
 		Msgf("revealed secret %x to %s", name, step.Current.GetIdentity())
 
 	return nil
 }
 
+// listAuditLogs implements commands. It performs the LIST_AUDIT_LOGS command
+func (c calypsoCommand) listAuditLogs(snap store.Snapshot, step execution.Step) error {
+	smcKey := step.Current.GetArg(SmcPublicKeyArg)
+	if len(smcKey) == 0 {
+		return xerrors.Errorf(notFoundInTxArg, SmcPublicKeyArg)
+	}
+
+	name := step.Current.GetArg(SecretNameArg)
+	if len(name) == 0 {
+		return xerrors.Errorf(notFoundInTxArg, SecretNameArg)
+	}
+
+	smcSecrets, ok := c.secrets[string(smcKey)]
+	if !ok {
+		return xerrors.Errorf(ErrorKeyNotFoundInSmcs, smcKey)
+	}
+
+	found := false
+	for _, s := range smcSecrets {
+		if s == string(name) {
+			found = true
+			break
+		}
+	}
```

Comment:
given how we use it, we should change the `c.secrets` type and make it instead a `secrets map[string]map[string]{}struct`, it would make it easier and more efficient to check for the presence or absence of a given key

Reply:
I'll look into it, but for now, smcSecrets is a list of strings, not sure how we can use a map[string]{}struct instead ?